### PR TITLE
Function regression

### PR DESCRIPTION
**Pull request by murphymadeleine21**
Robby's function from a few days ago in dataImport. Today's function for using elastic regression against effector function assays in regression.py

### CODE REVIEW

File: syserol/regression.py [hunk 1]
```diff
@@ -29,3 +29,38 @@ def patientComponents(nComp = 1):
...
+    func = importFunction()
+    igg = importIGG()
+    igg = igg.pivot(index='subject', columns = 'variable', values = 'value')
+    data_frames = [lum, glyc, func, igg]
+    df_merged = reduce(lambda  left,right: pd.merge(left,right,on=['subject'],
+                                                how='inner'), data_frames)
+    df_merged = df_merged.dropna()
+
+    #Z score
+    df_merged = df_merged.apply(zscore) ##Do we want to Zscore while function values are still in the dataframe?
```
Comment (aarmey) [inline syserol/regression.py]: With `normalize=True` ElasticNetCV will handle the normalization for you.

File: syserol/dataImport.py [hunk 2]
```diff
@@ -122,3 +122,16 @@ def createCube():
     assert ~np.any(np.all(np.isnan(cube), axis=(1, 2)))
 
     return cube, glyCube
+
+def funcRegImport():
+    df = importLuminex()
+    
+    df1 = df[df["variable"].str.contains("IgG1")].pivot(index="subject", columns="variable")
+    df2 = df[df["variable"].str.contains("IgG2")].pivot(index="subject", columns="variable")
```
Comment (aarmey) [inline syserol/dataImport.py]: Why do you have to separately pivot these?

**Review by aarmey — CHANGES_REQUESTED** [review]
Looks good—just a couple edits. Do you get roughly the same performance as in the paper?

File: syserol/dataImport.py (hunk 2 above)
Comment (murphymadeleine21) [reply syserol/dataImport.py]: Robby wrote this function so I'm not entirely sure. I think there were issues and this was the way he got it to work? I can request his review.

**Comment by murphymadeleine21** [comment]
Seems to be getting close values to the paper. ADCC was .603, ADCD .66. Those are the only two I have run so far.

**Comment by aarmey** [comment]
Great. That's reasonably close.

**Comment by murphymadeleine21** [comment]
Running the elastic net regression for functionality function with the updated code gives this repeated error:
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_coordinate_descent.py:529: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 0.24738483500540198, tolerance: 0.20337428878378377 

Not sure if this is because of one of the changes in the other files or something within sklearn? 
The function eventually runs but the prediction correlation went down a bit (ADCC .597)

**Comment by aarmey** [comment]
Try setting `max_iter=10000` in `ElasticNetCV`.

**Comment by murphymadeleine21** [comment]
Made this small figure function, not entirely sure I'm using all of the figure files correctly but this just plots predicted vs measured for any function we run through the elastic net regression.

**Comment by murphymadeleine21** [comment]
I've updated the figure file so it should be able to run now...
However, running the function took nearly an hour due to the regression function taking an extremely long time! Not sure why this is happening, and still getting the error with the convergence so might increase the max iterations even more?

File: syserol/regression.py [hunk 3]
```diff
@@ -28,3 +30,35 @@ def patientComponents(nComp=1):
...
+    #separate dataframes
+    df_func = df_merged[["ADCD", "ADCC", "ADNP", "CD107a", "IFNy", "MIP1b"]]
+    df_variables = df_merged.drop(['subject','ADCD', 'ADCC', 'ADNP', 'CD107a', 'IFNy', 'MIP1b'], axis = 1)
+
+    #perform regression
+    Y = df_func[function]
+    X = df_variables
+    Y_pred = np.empty(Y.shape)
+
+    Y_pred = cross_val_predict(ElasticNetCV(normalize=True, max_iter=10000), X, Y, cv=len(Y))
```
Comment (aarmey) [inline syserol/regression.py]: @murphymadeleine21 so the issue is we're doing cross-val inside of cross-val, so with ~180 subjects this results in fitting ~32,000 models. Instead of using `cross_val_predict`, `ElasticNetCV` has a member variable `alpha_` which gives the alpha it picked as best. You can then run `cross_val_predict` with `sklearn.linear_model.ElasticNet` and this best picked alpha to get the cross validation results. This should reduce the number of models fit to ~360. Finally, use `cv=10` instead of `cv=len(Y)`. This will then require 10 models fit instead of ~180 during the second half.

**Comment by murphymadeleine21** [comment]
This is also some older code that can hopefully be merged soon–we have figure 2 here which shows the predictability of our implemented Alter model.

File: syserol/dataImport.py [hunk 4]
```diff
@@ -132,3 +132,10 @@ def createCube():
     assert ~np.any(np.all(np.isnan(cube), axis=(1, 2)))
 
     return cube, glyCube
+
+def funcRegImport():
```
Comment (aarmey) [inline syserol/dataImport.py]: What does this function do? Are you using it? Delete if not.

**Review by aarmey — CHANGES_REQUESTED** [review]

File: syserol/dataImport.py (hunk 4 above)
Comment (murphymadeleine21) [reply syserol/dataImport.py]: Yeah I think this was Robby's function but I haven't been using it so I will delete.

**Review by aarmey — APPROVED** [review]
